File: backend/app/routers/analytics.py
"""API для аналитики: история тренировок, статистика по зонам."""

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy import func
from sqlalchemy.orm import Session

from ..database import get_db
from ..models import Athlete, HrReading, SessionAthlete, Session as TrainingSession
from ..schemas import AthleteStats, SessionStats, ZoneDistribution

router = APIRouter(prefix="/api/analytics", tags=["analytics"])


@router.get("/athletes/{athlete_id}/stats", response_model=AthleteStats)
def athlete_stats(athlete_id: str, db: Session = Depends(get_db)):
    """Возвращает агрегированную статистику спортсмена."""
    athlete = db.query(Athlete).filter(Athlete.id == athlete_id).first()
    if not athlete:
        raise HTTPException(404, "Спортсмен не найден")

    total_sessions = (
        db.query(func.count(func.distinct(HrReading.session_id)))
        .filter(HrReading.athlete_id == athlete_id, HrReading.session_id.isnot(None))
        .scalar() or 0
    )

    agg = (
        db.query(
            func.count(HrReading.id).label("count"),
            func.avg(HrReading.heart_rate).label("avg"),
            func.max(HrReading.heart_rate).label("max_hr"),
        )
        .filter(HrReading.athlete_id == athlete_id)
        .first()
    )

    duration = (
        db.query(func.sum(
            func.strftime("%s", SessionAthlete.left_at) - func.strftime("%s", SessionAthlete.joined_at)
        ))
        .filter(SessionAthlete.athlete_id == athlete_id, SessionAthlete.left_at.isnot(None))
        .scalar() or 0
    )

    return AthleteStats(
        total_sessions=total_sessions,
        total_duration_seconds=int(duration),
        avg_hr=round(agg.avg, 1) if agg and agg.avg else 0,
        max_hr_ever=agg.max_hr if agg and agg.max_hr else 0,
    )


@router.get("/athletes/{athlete_id}/history", response_model=list[SessionStats])
def athlete_history(athlete_id: str, limit: int = 20, db: Session = Depends(get_db)):
    """Возвращает историю тренировок спортсмена со статистикой по зонам."""
    athlete = db.query(Athlete).filter(Athlete.id == athlete_id).first()
    if not athlete:
        raise HTTPException(404, "Спортсмен не найден")

    readings = (
        db.query(
            HrReading.session_id,
            func.avg(HrReading.heart_rate).label("avg_hr"),
            func.max(HrReading.heart_rate).label("max_hr"),
            func.min(HrReading.heart_rate).label("min_hr"),
            func.count(HrReading.id).label("count"),
            func.sum(func.case((HrReading.zone == 1, 1), else_=0)).label("z1"),
            func.sum(func.case((HrReading.zone == 2, 1), else_=0)).label("z2"),
            func.sum(func.case((HrReading.zone == 3, 1), else_=0)).label("z3"),
            func.sum(func.case((HrReading.zone == 4, 1), else_=0)).label("z4"),
        )
        .filter(HrReading.athlete_id == athlete_id, HrReading.session_id.isnot(None))
        .group_by(HrReading.session_id)
        .order_by(HrReading.session_id.desc())
        .limit(limit)
        .all()
    )

    result = []
    for r in readings:
        session = db.query(TrainingSession).filter(TrainingSession.id == r.session_id).first()
        if not session:
            continue
        duration = 0
        if session.ended_at and session.started_at:
            duration = int((session.ended_at - session.started_at).total_seconds())
        result.append(SessionStats(
            session_id=r.session_id,
            session_name=session.name,
            avg_hr=round(r.avg_hr, 1) if r.avg_hr else 0,
            max_hr=r.max_hr or 0,
            min_hr=r.min_hr or 0,
            duration_seconds=duration,
            zones=ZoneDistribution(
                zone_1_seconds=int(r.z1 or 0),
                zone_2_seconds=int(r.z2 or 0),
                zone_3_seconds=int(r.z3 or 0),
                zone_4_seconds=int(r.z4 or 0),
            ),
        ))
    return result

File: backend-go/internal/handlers/analytics.go
package handlers

import (
	"database/sql"
	"net/http"
	"time"

	"github.com/go-chi/chi/v5"
)

// ── Analytics: агрегаты по hr_readings ────────────────────────

// AthleteStats: GET /api/analytics/athletes/{athlete_id}/stats
func (a *App) AthleteStats(w http.ResponseWriter, r *http.Request) {
	athleteID := chi.URLParam(r, "athlete_id")
	if a.getAthleteRow(w, athleteID) == nil {
		return
	}

	var totalSessions int
	_ = a.DB.QueryRow(
		`SELECT COUNT(DISTINCT session_id) FROM hr_readings
		 WHERE athlete_id = ? AND session_id IS NOT NULL`, athleteID,
	).Scan(&totalSessions)

	var (
		cnt   sql.NullInt64
		avgHr sql.NullFloat64
		maxHr sql.NullInt64
	)
	_ = a.DB.QueryRow(
		`SELECT COUNT(*), AVG(heart_rate), MAX(heart_rate) FROM hr_readings WHERE athlete_id = ?`,
		athleteID,
	).Scan(&cnt, &avgHr, &maxHr)

	var duration sql.NullFloat64
	_ = a.DB.QueryRow(
		`SELECT SUM(strftime('%s', left_at) - strftime('%s', joined_at))
		 FROM session_athletes WHERE athlete_id = ? AND left_at IS NOT NULL`, athleteID,
	).Scan(&duration)

	avg := 0.0
	if avgHr.Valid && avgHr.Float64 > 0 {
		avg = round1(avgHr.Float64)
	}
	maxHrEver := 0
	if maxHr.Valid {
		maxHrEver = int(maxHr.Int64)
	}
	totalDur := 0
	if duration.Valid {
		totalDur = int(duration.Float64)
	}

	writeJSON(w, 200, map[string]any{
		"total_sessions":         totalSessions,
		"total_duration_seconds": totalDur,
		"avg_hr":                 avg,
		"max_hr_ever":            maxHrEver,
	})
}

// AthleteHistory: GET /api/analytics/athletes/{athlete_id}/history?limit=20
func (a *App) AthleteHistory(w http.ResponseWriter, r *http.Request) {
	athleteID := chi.URLParam(r, "athlete_id")
	if a.getAthleteRow(w, athleteID) == nil {
		return
	}
	limit := intQueryParam(r, "limit", 20)

	// Важно: MaxOpenConns(1) — сначала вычитываем агрегаты и закрываем rows,
	// только потом делаем запросы сессий (иначе дедлок на единственном коннекте).
	type aggRow struct {
		sessionID      string
		avgHr          sql.NullFloat64
		maxHr, minHr   sql.NullInt64
		z1, z2, z3, z4 sql.NullInt64
	}
	aggs := []aggRow{}
	{
		rows, err := a.DB.Query(`
			SELECT session_id, AVG(heart_rate), MAX(heart_rate), MIN(heart_rate),
			       SUM(CASE WHEN zone = 1 THEN 1 ELSE 0 END),
			       SUM(CASE WHEN zone = 2 THEN 1 ELSE 0 END),
			       SUM(CASE WHEN zone = 3 THEN 1 ELSE 0 END),
			       SUM(CASE WHEN zone = 4 THEN 1 ELSE 0 END)
			FROM hr_readings
			WHERE athlete_id = ? AND session_id IS NOT NULL
			GROUP BY session_id
			ORDER BY session_id DESC
			LIMIT ?`, athleteID, limit,
		)
		if err != nil {
			httpError(w, 500, err.Error())
			return
		}
		for rows.Next() {
			var g aggRow
			if err := rows.Scan(&g.sessionID, &g.avgHr, &g.maxHr, &g.minHr, &g.z1, &g.z2, &g.z3, &g.z4); err != nil {
				_ = rows.Close()
				httpError(w, 500, err.Error())
				return
			}
			aggs = append(aggs, g)
		}
		_ = rows.Close()
	}

	out := []map[string]any{}
	for _, g := range aggs {
		var (
			sessName  sql.NullString
			startedAt sql.NullString
			endedAt   sql.NullString
		)
		err := a.DB.QueryRow(`SELECT name, started_at, ended_at FROM sessions WHERE id = ?`, g.sessionID).
			Scan(&sessName, &startedAt, &endedAt)
		if err == sql.ErrNoRows {
			continue
		}
		if err != nil {
			httpError(w, 500, err.Error())
			return
		}

		duration := 0
		if endedAt.Valid && startedAt.Valid {
			st := parseDBTime(startedAt.String)
			en := parseDBTime(endedAt.String)
			if !st.IsZero() && !en.IsZero() {
				duration = int(en.Sub(st).Seconds())
			}
		}

		avg := 0.0
		if g.avgHr.Valid && g.avgHr.Float64 > 0 {
			avg = round1(g.avgHr.Float64)
		}

		out = append(out, map[string]any{
			"session_id":       g.sessionID,
			"session_name":     nullStr(sessName),
			"avg_hr":           avg,
			"max_hr":           nullInt64(g.maxHr),
			"min_hr":           nullInt64(g.minHr),
			"duration_seconds": duration,
			"zones": map[string]int{
				"zone_1_seconds": nullInt0(g.z1),
				"zone_2_seconds": nullInt0(g.z2),
				"zone_3_seconds": nullInt0(g.z3),
				"zone_4_seconds": nullInt0(g.z4),
			},
		})
	}
	writeJSON(w, 200, out)
}

func parseDBTime(s string) time.Time {
	if t, err := time.Parse(time.RFC3339Nano, s); err == nil {
		return t
	}
	for _, layout := range []string{"2006-01-02 15:04:05.999999", "2006-01-02 15:04:05"} {
		if t, err := time.ParseInLocation(layout, s, time.UTC); err == nil {
			return t
		}
	}
	return time.Time{}
}
